src/hotspot/share/opto/graphKit.cpp

```diff
@@ -3357,7 +3358,15 @@ Node* GraphKit::gen_checkcast(Node *obj, Node* superklass,
 
   if (cast_obj == nullptr) {
     // Generate the subtype check
-    Node* not_subtype_ctrl = gen_subtype_check(not_null_obj, superklass );
+    Node* improved_superklass = superklass;
+    if (improved_klass_ptr_type != klass_ptr_type) {
+      if (improved_klass_ptr_type->singleton()) {
+        improved_superklass = makecon(improved_klass_ptr_type);
```

Comment:
Do you really need to special case that one? Wouldn't the CastPP constant fold if `improved_klass_ptr_type` is a singleton?
Also there is the question of whether a `CastPP` or a `CheckCastPP` should be used here and if it matters. They don't differ that much anymore so I suppose `CastPP` is ok.

Reply:
> Do you really need to special case that one? Wouldn't the CastPP constant fold if improved_klass_ptr_type is a singleton?

For `Y <: abstract X`, when `superklass` is a precise constant `X` and `improved_klass_ptr_type` is a precise constant `Y`, then `CastPP(nullptr, superklass, improved_klass_ptr_type)` will be replaced by `top` and assigned to `improved_superklass` which is probably not what we want. That's why I special cased the singleton case. This could happen, for example, with:
```
obj = (X)o; // superklass = ConP #precise X, improved_klass_ptr_type = precise Y
```

> Also there is the question of whether a CastPP or a CheckCastPP should be used here and if it matters. They don't differ that much anymore so I suppose CastPP is ok.

Okay, sound good.

Reply:
Ok. Makes sense. Thanks for the explanation.

Reply:
Then isn't there a risk that after some transformation the `CastPP` ends up with an input that's a constant superklass which would cause the `CastPP` to transform to top?

Reply:
Hm, you're right. That's the very same problem. So, a `CastPP` does not really work here. Should we go back to the previously suggested version without `CastPP`?
```
    if (improved_klass_ptr_type != klass_ptr_type) {
      if (improved_klass_ptr_type->singleton()) {
        improved_superklass = makecon(improved_klass_ptr_type);
      } else {
        superklass->raise_bottom_type(improved_klass_ptr_type);
        _gvn.set_type(superklass, improved_klass_ptr_type);
      }
    }
```
It's a best effort solution. We might still miss opportunities to remove sub type checks later where we could call `try_improve()` again. But at least the `SubTypeCheck` is now in sync with the `CheckCastPP` that is also only improved at parse time with `try_improve()` and not later anymore.

We could think about using `try_improve()` during IGVN as well to get better type information later. But I suggest to do that separately to this fix.

Reply:
> Should we go back to the previously suggested version without CastPP?

Isn't there a risk with that one too? if `superklass` is a `TypeNode` and its input changes to a constant for instance.

Reply:
If doing this in `GraphKit` doesn't work well, maybe it should be done in `SubTypeCheckNode::Value`? This way it would apply all all stage of compilation? (That assumes we don't care of what happens if `ExpandSubTypeCheckAtParseTime` is true).

Reply:
> > Should we go back to the previously suggested version without CastPP?
> 
> Isn't there a risk with that one too? if `superklass` is a `TypeNode` and its input changes to a constant for instance.

I'm afraid you're right. This could probably happen, too.

> If doing this in `GraphKit` doesn't work well, maybe it should be done in `SubTypeCheckNode::Value`? This way it would apply all all stages of compilation? (That assumes we don't care of what happens if `ExpandSubTypeCheckAtParseTime` is true).

I've first thought of doing it in `SubTypeCheckNode::Value()` but assumed we can get away with handling it in `GraphKit`. But as now figured out, this comes with new problems and does not seem to be safe. I will try to undo my current fix idea in `GraphKit` and do it in `SubTypeCheckNode::Value()` instead. This should work.

Reply:
I could not find a test that shows a benefit by doing the improved check with code from `try_improve()` inside `SubTypeCheckNode::Value()`. The original patch only showed a win for mainline when we have two identical `SubTypeCheckNodes` such that they can common up with an improved constant. But this cannot be achieved when having the code in `SubTypeCheckNode::Value()`.

I therefore suggest to revert back to the original version to directly plug in a better constant, if we find one with `try_improve()`, and just skip the other non-constant cases with `LoadKlass` etc.

For the Valhalla bug, I can do the more sophisticated fix to improve `SubTypeCheckNode::Value()` with code from `try_improve()` but it does not seem worth for mainline. What do you think?

Reply:
Sounds good. Thanks for giving it a try.
Since you mention fixing this in valhalla, do you expect it makes a difference there but not in mainline?

Reply:
Ok great! I had another look at Valhalla and I think the constant improvement fix is all that is needed. I reverted the last commit to go back to the original patch.